### ISSUE TITLE
make eeprom layers not shadow the progmem layers

### ISSUE THREAD

**Issue by obra** [keyboardio/Kaleidoscope-EEPROM-Keymap#15]
oh. @algernon I was thinking - rather than having eeprom layers as overlays on top of the core layers, why not have them as a separate set of layers and store our 'boot layer' id in eeprom?
it would simplify implementation AND get more potential layers
algernonToday at 12:57 PM
that's certainly a possibility
can you open an issue against EEPROM-Keymap? I'll do the magic tomorrow then. (it's just another getKey implementation, and perhaps a tiny bit of eeprom tweaking)
obraToday at 12:58 PM
sure.

**Comment by algernon** [comment]
I was thinking about splitting this in two: one part would be here, that instead of overlaying, extends the layers; another part would go to another - possibly new - plugin, that simply stores whether we are on PROGMEM or EEPROM layers (I wouldn't store the exact layer, because if you reboot while on the `Fn` layer for example, you likely don't want to end up there after the keyboard boots up again).

...now that I spelled it out, this could be part of EEPROMKeymap, as an optional extra plugin (so we'd provide two objects). The main reason I don't want to just add a boot id byte is because it's not needed for neither EEPROM-only, nor the overlay, and that one byte of EEPROM is precious. On the other hand, always having a byte there means its easier to switch between modes.

**Comment by algernon** [comment]
In any case, while I ponder about that, there's another thing that came to mind last night while thinking about this: I'd like to make the interface easier. Currently, overlay is the default, but changing that is not intuitive. Nor is setting up the layers we override. So without much further ado:

```c++
// Sets Layer.getKey to getKeyOverride, and max_layers(4)
EEPROMKeymap.setup(EEPROMKeymap::OVERLAY, 4);

// Sets Layer.getKey to getKey, and max_layers(5)
EEPROMKeymap.setup(EEPROMKeymap::DIRECT, 5);

// Sets Layer.getKey to getKeyExtended, and max_layers(5)
EEPROMKeymap.setup(EEPROMKeymap::EXTEND, 5);
```

**Comment by algernon** [comment]
Ok, so I have this down *mostly*, with one piece missing: saving the "default" layer. I have a question about how that should work. If we have a PROGMEM keymap, and extend it with additional layers in EEPROM, that can work wonderfully, indeed. But how do you switch to the new layers? We'd either have to prepare the PROGMEM keymap to be able to switch to layers higher than in PROGMEM, or we'd need a Focus hook, or a Leader sequence, MagicCombo, whathaveyou.

What I'm trying to say is that this PROGMEM+EEPROM extension mechanism can work fine, if the firmware is prepared for that mode of operation. And in that case, saving the default keymap should be outside `EEPROMKeymap`, because one may very well wish to implement a logic that makes sense for THEIR setup.

**Comment by algernon** [comment]
The [f/extend](https://github.com/keyboardio/Kaleidoscope-EEPROM-Keymap/tree/f/extend) branch has a work in progress implementation. I need to figure out what to do with `keymap.map`, need to teach that about extending too.

Should likely provide a way to figure out how many progmem layers there are, and which can be edited. Neither of this is a problem for the `DIRECT` and `OVERLAY` modes, but its essential for `EXTEND` to work.

Most likely, I'll just store an `uint32_t` and if in EXTEND mode, flip the bits for the PROGMEM layers. Then make this queryable (but not settable) via `keymap.roLayers`. Or, to make it simpler, I could just store the mode, and provide a way to check the value of `progmem_layers_`. Not sure about the name of that command. As a third option, I could store the mode, and still provide `keymap.roLayers`, and compute that on the fly. This will allow us later on to change the internals without breaking anything user-facing.

As for why would one want a bitmap for editable layers? Because it is possible to have a set of layers in PROGMEM, another set in EEPROM, and reorder them in custom ways. Mind you, I'm not sure we want to support such an edge-case right now.

In conclusion:
- [ ] store the mode within the object
- [ ] make it possible to query the id of the first read-write layer / number of RO layers

This assumes that EEPROM layers are put after PROGMEM layers (which is the case now with EXTEND). But its also possible to put them before, in which case we'd need a different property, or perhaps a read-write range. Or we just don't care about this scenario just yet, and focus on what's actually useful and desirable.

**Comment by algernon** [comment]
With the idea "explained" above, `keymap.map` would always return the whole keymap, but only update the EEPROM parts. A new command, `keymap.roLayers` would return the number of PROGMEM layers, so UIs can still show those layers, but disable editing.

**Comment by algernon** [comment]
The `f/extend` branch mentioned above now supports the `EXTEND` method, and `keymap.map` works as it should (ignoring the first `progmem_layers_` layers). I still need to implement the `keymap.roLayers` Focus command, clean up the branch, and will submit a PR then.

It won't have the store-the-layer-id-in-eeprom parts, we should discuss that separately, I think.

**Comment by algernon** [comment]
With #17 and keyboardio/Kaleidoscope-EEPROM-Settings#9 merged, I think this can be considered done.
